router/http.go

```diff
@@ -356,310 +345,27 @@ type httpService struct {
-		return
-	}
-
-	isSticky := false
-	stickyAddr := ""
-	if req.Header.Get(hdrUseStickySessions) == "true" {
-		// TODO(bgentry): switch to better way to check sticky setting
-		isSticky = true
-		if stickyAddr = s.stickyCookieAddr(req); stickyAddr != "" {
-			sortStringFirst(addrs, stickyAddr)
-		}
-	}
-	req.Header.Del(hdrUseStickySessions) // delete this no matter what
-
-	// Most of this is borrowed from httputil.ReverseProxy
-	outreq := &http.Request{}
-	*outreq = *req // includes shallow copies of maps, but okay
-
 	// Pass the Request-URI verbatim without any modifications
-	outreq.URL.Opaque = strings.Split(strings.TrimPrefix(req.RequestURI, req.URL.Scheme+":"), "?")[0]
-	outreq.URL.Scheme = "http"
-	outreq.Proto = "HTTP/1.1"
-	outreq.ProtoMajor = 1
-	outreq.ProtoMinor = 1
-	outreq.Close = false
-
-	// TODO: Proxy HTTP CONNECT? (example: Go RPC over HTTP)
-
-	// Directly bridge `Connection: Upgrade` requests
-	if strings.ToLower(outreq.Header.Get("Connection")) == "upgrade" {
-		s.forwardAndProxyTCP(w, outreq, addrs, stickyAddr, isSticky)
-		return
-	}
+	req.URL.Opaque = strings.Split(strings.TrimPrefix(req.RequestURI, req.URL.Scheme+":"), "?")[0]
 
-	// Remove hop-by-hop headers to the backend.  Especially
-	// important is "Connection" because we want a persistent
-	// connection, regardless of what the client sent to us.  This
-	// is modifying the same underlying map from req (shallow
-	// copied above) so we only copy it if necessary.
-	copiedHeaders := false
-	for _, h := range hopHeaders {
-		if outreq.Header.Get(h) != "" {
-			if !copiedHeaders {
-				outreq.Header = make(http.Header)
-				copyHeader(outreq.Header, req.Header)
-				copiedHeaders = true
-			}
-			outreq.Header.Del(h)
-		}
-	}
-
-	var (
-		backend string
-		res     *http.Response
-		err     error
-	)
-
-	for _, backend = range addrs {
-		// TODO: limit number of backends tried
-		// TODO: temporarily quarantine failing backends
+	//TODO(benburkert): only call this when the ss set is updated, instead of every request
+	s.proxy.UpdateBackends(s.sc.Addrs())
 
-		outreq.URL.Host = backend
-		res, err = transport.RoundTrip(outreq)
-		if err != nil {
-			if _, ok := err.(dialErr); ok {
-				// retry, maybe log a message about it
-				continue
-			}
-			log.Println("http: proxy error:", err)
-			fail(w, 503)
-			return
-		}
-		defer res.Body.Close()
-		break
-	}
-
-	if res == nil {
-		log.Println("no backends available")
-		fail(w, 503)
-		return
-	}
-
-	if isSticky && stickyAddr != backend {
-		http.SetCookie(w, s.newStickyCookie(backend))
-	}
-
-	for _, h := range hopHeaders {
-		res.Header.Del(h)
-	}
-
-	copyHeader(w.Header(), res.Header)
-
-	w.WriteHeader(res.StatusCode)
-	w.(http.Flusher).Flush()
-	_, err = io.Copy(httphelper.FlushWriter{Writer: w, Enabled: true}, res.Body) // TODO(bgentry): consider using a flush interval
-	if err != nil {
-		log.Println("reverse proxy copy err:", err)
-		return
+	if isWebSocketUpgrade(req) {
+		s.proxy.ServeWebSocket(w, req)
```

Comment:
I still don't think that this interface should be exposed to users of the `Proxy` type. These should be internal implementation details for that type.

Why not stick with the `http.Handler` interface and let the implementation of `Proxy` choose to be smart enough to handle conn upgrades?